gcp/src/main/java/org/apache/iceberg/gcp/biglake/BigLakeCatalog.java

```diff
@@ -0,0 +1,396 @@
+      // (https://github.com/googleapis/google-cloud-java#application-default-credentials)
+      // that works on GCP services (e.g., GCE, GKE, Dataproc).
+      initClient =
+          new BigLakeClient(
+              initProperties.getOrDefault(CatalogProperties.URI, DEFAULT_BIGLAKE_SERVICE_ENDPOINT));
+    } catch (IOException e) {
+      throw new ServiceFailureException(e, "Creating BigLake client failed");
+    }
+
+    initialize(initName, initProperties, initClient);
+  }
+
+  @VisibleForTesting
+  void initialize(String initName, Map<String, String> initProperties, BigLakeClient initClient) {
+    this.name = initName;
+    this.properties = ImmutableMap.copyOf(initProperties);
+
+    Preconditions.checkArgument(initClient != null, "BigLake client must not be null");
+    this.client = initClient;
+
+    Preconditions.checkArgument(
+        properties.containsKey(GCPProperties.PROJECT_ID), "GCP project ID must be specified");
+    this.projectId = properties.get(GCPProperties.PROJECT_ID);
+
+    Preconditions.checkArgument(
+        properties.containsKey(GCPProperties.REGION), "GCP region must be specified");
+    this.region = properties.get(GCPProperties.REGION);
+
+    // Users can specify the BigLake catalog ID, otherwise catalog plugin name will be used.
+    // For example, "spark.sql.catalog.<plugin-name>=org.apache.iceberg.spark.SparkCatalog"
+    // specifies the plugin name "<plugin-name>".
+    this.catalogId = properties.getOrDefault(GCPProperties.BIGLAKE_CATALOG_ID, initName);
+    this.catalogName = CatalogName.of(projectId, region, catalogId);
+
+    Preconditions.checkArgument(
+        properties.containsKey(CatalogProperties.WAREHOUSE_LOCATION),
+        "Data warehouse location must be specified");
+    this.warehouseLocation =
+        LocationUtil.stripTrailingSlash(properties.get(CatalogProperties.WAREHOUSE_LOCATION));
+
+    String ioImpl =
+        properties.getOrDefault(CatalogProperties.FILE_IO_IMPL, ResolvingFileIO.class.getName());
+    this.io = CatalogUtil.loadFileIO(ioImpl, properties, conf);
+
+    this.closeableGroup = new CloseableGroup();
+    closeableGroup.addCloseable(client);
+    closeableGroup.addCloseable(io);
+    closeableGroup.setSuppressCloseFailure(true);
+  }
+
+  @Override
+  protected TableOperations newTableOps(TableIdentifier identifier) {
+    String dbId = databaseId(identifier.namespace());
+    validateDatabaseId(dbId, identifier.namespace());
+
+    return new BigLakeTableOperations(client, io, name(), tableName(dbId, identifier.name()));
+  }
+
+  @Override
+  protected String defaultWarehouseLocation(TableIdentifier identifier) {
+    String dbId = databaseId(identifier.namespace());
+    validateDatabaseId(dbId, identifier.namespace());
+
+    String locationUri = loadDatabase(dbId).getHiveOptions().getLocationUri();
+    return String.format(
+        "%s/%s",
+        Strings.isNullOrEmpty(locationUri) ? databaseLocation(dbId) : locationUri,
```

Comment:
I can see that the intent here is for `databaseLocation` to be the default if the location is not explicitly set in metadata, but this naming is confusing. I would expect `databaseLocation(dbId)` to always produce the correct location (whether stored in the metastore or not) because that's what the name implies.

I think this should either rename `databaseLocation` to something more `defaultWarehouseLocation(Namespace)` or `newLocation(databaseId)`. If you use this naming, then move `loadDatabase(dbId).getHiveOptions().getLocationUri()` into that method.

Reply:
Renamed it to newDbLocation, because databaseLocation is used at another place for constructing URI of a new DB,  something like newDbLocation is needed anyway.